src/vcf/validator.cpp

```diff
@@ -122,13 +126,63 @@ namespace ebi
     }
 
     bool is_valid_vcf_file(std::istream &input,
-                           ebi::vcf::Parser &validator,
+                           const std::string &sourceName,
+                           ValidationLevel validationLevel,
+                           Ploidy ploidy,
                            std::vector<std::unique_ptr<ebi::vcf::ReportWriter>> &outputs)
     {
+        std::vector<char> line;
+        ebi::util::readline(input, line);
+        ebi::vcf::Version version;
+        try {
+            version = detect_version(line);
+        } catch (FileformatError * error) {
+            for (auto &output : outputs) {
+                output->write_error(*error);
+            }
+            return false;
+        }
+        std::unique_ptr<Parser> validator = build_parser(sourceName, validationLevel, version, ploidy);
+        return validate(line, input, *validator, outputs);
+    }
 
+    Version detect_version(const std::vector<char> &line)
```

Comment:
several points here:
- yes, it's a good idea to get rid of the `compare`.
- yes, one throw makes it cleaner
- not sure about substr (it mixes positions and counts, in fact the parameter should be 13 in both usages, which is unintuitive), I tried to avoid on purpose that kind of index harcoding.

what about:

```c++
        std::string common_substring{"##fileformat="};
        std::vector<char>::const_iterator lineIt;
        std::string::const_iterator commonIt;
        std::tie(commonIt, lineIt) = std::mismatch(common_substring.begin(), common_substring.end(), line.begin());

        if (commonIt == common_substring.end()) {
            std::string provided_version{lineIt, line.end()};
            util::remove_end_of_line(provided_version);

            if (provided_version == "VCFv4.1") {
                return Version::v41;
            } else if (provided_version == "VCFv4.2") {
                return Version::v42;
            } else if (provided_version == "VCFv4.3") {
                return Version::v43;
            }
        }

        throw new FileformatError{1, "The fileformat declaration is not valid (the line must start with "
                    + common_substring + " and the value must be one of 'VCFv4.1', 'VCFv4.2' or 'VCFv4.3')"};
```

if you really really want to remove that `std::mismatch`, we could do:
```c++
        std::string common_substring{"##fileformat="};
        std::string line{vector_line.begin(), vector_line.end()};

        if (line.substr(0, common_substring.size()) == common_substring) {
            std::string provided_version{line.substr(common_substring.size())};
            util::remove_end_of_line(provided_version);

            if (provided_version == "VCFv4.1") {
                return Version::v41;
            } else if (provided_version == "VCFv4.2") {
                return Version::v42;
            } else if (provided_version == "VCFv4.3") {
                return Version::v43;
            }
        }

        throw new FileformatError{1, "The fileformat declaration is not valid (the line must start with "
                    + common_substring + " and the value must be one of 'VCFv4.1', 'VCFv4.2' or 'VCFv4.3')"};
```